Plugin telemetry events

akoutmos opened the pull request:
This PR addresses some of my thoughts that I raised in #892 and emits start+stop events for dataloader events where there are `pending_batches` to process. Those two additional Telemetry events are:
`[:absinthe, :dataloader, :resolve, :start]` when a dataloader with pending batches starts
`[:absinthe, :dataloader, :resolve, :stop]` when a dataloader with pending batches finishes

This is just an initial stab at the problem, so feel free to comment. Thanks :)!

benwilson512 commented:
Hey @akoutmos thanks for taking a stab at this!

Here's the challenge we need to solve: Dataloader is an _optional_ dependency within Absinthe. I don't think it's appropriate for Absinthe to emit dataloader specific events. I _do_ think that the resolution phase should emit events for when a plugin callback has started, and when a plugin callback has ended though, and the plugin module would be part of the event emitted.

For Dataloader specific events, I think we need to have those fire within Dataloader itself.

benwilson512 commented on lib/absinthe/middleware/dataloader.ex:
```diff
@@ -3,15 +3,46 @@ if Code.ensure_loaded?(Dataloader) do
+
+          emit_start_event(start_time, exec, loader)
+          updated_loader = Dataloader.run(loader)
+          emit_stop_event(start_time, exec, loader)
```
Specifically, I think we should move these start and end events into https://github.com/absinthe-graphql/absinthe/blob/master/lib/absinthe/phase/document/execution/resolution.ex#L78 and it would be something like:

```
  defp run_callbacks(plugins, callback, acc, true) do
    Enum.reduce(plugins, acc, fn plugin, acc ->
      start_time = System.monotonic_time()
              :telemetry.execute(
          [:absinthe, :plugin, callback, :start]
          %{start_time: start_time},
          %{plugin: plugin, acc: acc}
        )
      acc = apply(plugin, callback, [acc])
      # stop event
     acc
    end)
  end
```

akoutmos commented:
@benwilson512 I think my latest commits address the comments that you brought up. I really like this approach compared to my original solution as it is more general purpose now and covers all the plugins. On that same note, is a dataloader specific event even needed? You can match on 
```elixir
{
  :telemetry_event, 
  [:absinthe, :plugin, :callback, :start], 
  %{start_time: _}, 
  %{loader:Absinthe.Middleware.Dataloader}
}
```
to get your dataloader (or any plugin for that matter) specific events. Thoughts?

benwilson512 commented on lib/absinthe/phase/document/execution/resolution.ex:
```diff
@@ -75,11 +78,35 @@ defmodule Absinthe.Phase.Document.Execution.Resolution do
+    Enum.reduce(plugins, acc, fn plugin, acc ->
+      start_time = System.monotonic_time()
+
+      emit_start_event(start_time, acc, plugin)
```
Very close! So, `callback` here is either the atom `:before_resolution` or `:after_resolution`, and I was thinking we could use that to dispatch more finely tuned events. Specifically this would get us:

```
[:absinthe, :plugin, :before_resolution, :start]
[:absinthe, :plugin, :before_resolution, :stop]
[:absinthe, :plugin, :after_resolution, :start]
[:absinthe, :plugin, :after_resolution, :stop]
```

benwilson512 commented on lib/absinthe/phase/document/execution/resolution.ex:
```diff
@@ -75,11 +78,35 @@ defmodule Absinthe.Phase.Document.Execution.Resolution do
+    :telemetry.execute(
+      @callback_stop_event,
+      %{duration: System.monotonic_time() - start_time},
+      %{acc: acc, loader: loader}
```
The `loader` variable here is I think the `plugin` right? We probably want to call it `plugin: plugin`

akoutmos commented:
My bad. I read your callback line as an atom and not the function variable. Third time is a charm I think I addressed all your feedback :D. Thoughts on the separate dataloader telemetry events? Thanks again!!

binaryseed commented on lib/absinthe/phase/document/execution/resolution.ex:
```diff
@@ -75,11 +75,35 @@ defmodule Absinthe.Phase.Document.Execution.Resolution do
+  defp emit_start_event(start_time, acc, callback, plugin) do
+    :telemetry.execute(
+      [:absinthe, :plugin, callback, :start],
+      %{start_time: start_time},
```
For this to be a useful measurement, it needs to be a `system_time()`... `monotonic_time` is still needed for duration tracking, but it doesn't actually mark a point in time

akoutmos replied:
Updated the PR to have the `start_time` be from system_time()

binaryseed commented on test/absinthe/middleware/dataloader_test.exs:
```diff
@@ -158,11 +159,57 @@ defmodule Absinthe.Middleware.DataloaderTest do
+    # This test emits a total of 24 events, but there we are only validating the
+    # dataloader events in this test
+    results =
+      Enum.reduce(1..24, %{dataloader_starts: 0, dataloader_stops: 0}, fn _, acc ->
```
I'm unclear why the `reduce` is needed, the other telemetry tests managed to work just fine with `assert_receive`...

https://github.com/absinthe-graphql/absinthe/blob/7821d278b8c511fbae752f5f68bd94195a9c2607/test/absinthe/integration/execution/telemetry_test.exs#L62

akoutmos replied:
The reason for the reduce is so that I can count only the specific events that I want to fetch. Since the telemetry events emit on all plugins, I was also getting Async and Batch events. But only wanted to assert on the Dataloader specific events.

binaryseed commented on test/absinthe/middleware/dataloader_test.exs:
```diff
@@ -158,11 +159,57 @@ defmodule Absinthe.Middleware.DataloaderTest do
 
+    # Get test PID and clear mailbox
+    self = self()
+    flush_process_mailbox()
```
Why do you need to do this?

akoutmos replied:
Fixed root cause of test flakiness by making tests non-async. Deleted this block

binaryseed commented:
Also, we have a pattern of sticking an `id` on each event to correlate the start & stop...

https://github.com/absinthe-graphql/absinthe/blob/7821d278b8c511fbae752f5f68bd94195a9c2607/lib/absinthe/middleware/telemetry.ex#L12

benwilson512 commented:
Is this ready for re-review?

akoutmos commented:
I think I addressed everything on my end. The only open question from what I gather is the number of events being generated...but I defer to you guys on that one.

binaryseed commented:
So, I looked into this locally on my API, and the problem is real.

We've got a few plugins installed for standard stuff like auth & monitoring, and this addition calls `execute` 2 times for every plugin for every phase of resolution.

That means when using stuff like `Async` / `Batch`, every new phase added by any plugin re-triggers a whole new set of telemetry events for every plugin.

For example, I tried it out with a really simple query, and there were 6 resolution phases. Each phase added 14 calls to execute (we have 7 plugins). That means a total of 84 additional calls to `telemetry.execute` - 84 more ETS lookups. And that's as a no-op, if there's a handler attached, then it's going to get called 84 times.

I'm a little worried about this... It's pretty unlikely that the raw data here is going to be leveraged, and that's a good bit of work to do for something that's rarely going to be used.

I"m wondering if directly instrumenting Dataloader would be a better way to go about this vs instrumenting every single plugin interaction?

benwilson512 commented:
@binaryseed This is a good point, I didn't think about the fact that this will run N*M times, where N is the number of plugins, and M is the number of resolution phases that happen (roughly proportional to the depth of the GraphQL query). `:ets` is pretty fast, but I agree that there isn't a lot of point to firing events about plugins that don't execute.

Perhaps an alternative strategy would be to emit events only within the functions within the `batch` and `dataloader` plugins that actually execute user code so that you can track how long a given `batch` or `run_batch` call takes if any are queued up without firing events at all if none are present. I realize this is closer to what you had originally @akoutmos and I apologize for the back and forth here. In this proposal to be clear, we would move the event hooks to here: https://github.com/absinthe-graphql/absinthe/blob/master/lib/absinthe/middleware/batch.ex#L156 and within Dataloader itself. @binaryseed does that seem more reasonable?

akoutmos commented:
No worries at all @benwilson512! That sounds good to me in regards to moving the telemetry events to the plugins directly. If that sounds good to you as well @binaryseed then I can get going on this approach and update this PR + open up a new one for the dataloader repo.

binaryseed commented:
Instrumenting `call_batch_fun` sounds pretty good!

benwilson512 commented:
@akoutmos do note the naming conventions mentioned here: https://github.com/absinthe-graphql/absinthe/pull/901#pullrequestreview-386405817

akoutmos commented:
Working on cleaning up this PR to reflect the conversation that we had. But managed to wrap up the Dataloader changes: https://github.com/absinthe-graphql/dataloader/pull/89

Will update this PR over the weekend. Thanks!

akoutmos commented:
I think I addressed everything that we discussed and also updated the documentation.

Question for you guys. Should I remove the telemetry tests in the `test/absinthe/middleware/dataloader_test.exs` file? The emitting of telemetry events is covered in the Dataloader PR that I made...but it would also be nice to ensure that those events are captured in an integration style test within Absinthe. On the flip side, the tests in Absinthe will fail until a newer version of Dataloader is published and the deps are updated in Absinthe. Thoughts?

binaryseed commented:
Yeah, I'd much rather we keep the tests for dataloader telemetry events inside the library itself...

akoutmos commented:
I went ahead and removed the test @binaryseed

binaryseed commented:
Thanks! I'm going to merge this and followup with a small commit to move the `telemetry.execute` call in the start case to execute after the task is spawned so that handlers don't impact the timing!

binaryseed commented:
Here it is: https://github.com/absinthe-graphql/absinthe/commit/38016db1216bcd133273dd6ff46885cb06611820

akoutmos commented:
Thanks @binaryseed! Let me know what you think of the dataloader changes and if you require any changes there https://github.com/absinthe-graphql/dataloader/pull/89.
